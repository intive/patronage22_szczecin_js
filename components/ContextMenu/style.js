import styled from 'styled-components'

export const StyledUl = styled.ul`
  margin: 0;
  padding: 0;
`

export const StyledContextMenu = styled.div`
  position: relative;
  text-align: right;
`

export const Button = styled.button`
  border: none;
  cursor: pointer;
  padding: 0;
  color: #7e7e7e;
  border-radius: 4px;
  margin-bottom: 0.25rem;
  line-height: 0.5;
  background: none;

  &:hover, &[aria-expanded="true"] {
    background: #f7f7f7;
  }
`

export const Nav = styled.nav`
  display: none;
  position: absolute;
  padding: 0.5rem 0rem;
  background: #ffffff;
  box-shadow: 0px 1px 2px rgba(98, 98, 98, 0.24);
  border-radius: 4px;
  right: 0px;

  &.active {
    display: block;
  }
`
