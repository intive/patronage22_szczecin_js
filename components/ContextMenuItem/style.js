import styled from 'styled-components'
import Icon from '../Icon/Icon'

export const StyledLi = styled.li`
  list-style: none;
  margin: 0;
  padding: 0;

  &:last-child button {
    margin-bottom: 0;
  }
`

export const ContextMenuButton = styled.button`
  display: flex;
  align-items: center;
  background: none;
  line-height: 1.5rem;
  color: #626262;
  border: none;
  cursor: pointer;
  padding: 0.25rem 0.5rem;
  width: 100%;
  font-size: 0.875rem;
  white-space: nowrap;

  &:hover {
    background: #bababa9e;
  }
`
export const StyledIcon = styled(Icon)`
  width: 24px;
  height: 24px;
  margin-right: 0.25rem;
  color: #7e7e7e;
`
