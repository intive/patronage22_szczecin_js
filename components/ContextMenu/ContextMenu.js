import { useState, useRef } from 'react'
import Icon from '../Icon/Icon'
import { StyledContextMenu, Button, Nav, StyledUl } from './style'
import useClickOutside from '../../hooks/useClickOutside'

export default function ContextMenu (props) {
  const [isActive, setIsActive] = useState(false)
  const ref = useRef()

  useClickOutside(ref, () => setIsActive(false))

  const changeStateHandler = (e) => {
    e.preventDefault()
    setIsActive(!isActive)
  }

  return (
    <StyledContextMenu ref={ref}>
      <Button onClick={changeStateHandler} aria-haspopup='menu' aria-expanded={isActive} aria-controls={props.id}>
        <Icon name='more_horiz' />
      </Button>

      <Nav onClick={changeStateHandler} id={props.id} className={isActive && 'active'}>
        <StyledUl>{props.children}</StyledUl>
      </Nav>
    </StyledContextMenu>
  )
}
