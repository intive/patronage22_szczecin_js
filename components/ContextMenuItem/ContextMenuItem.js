import { ContextMenuButton, StyledIcon, StyledLi } from './style'

export default function ContextMenuItem (props) {
  return (
    <StyledLi>
      <ContextMenuButton onClick={props.onClick}>
        <StyledIcon name={props.icon} />
        {props.name}
      </ContextMenuButton>
    </StyledLi>
  )
}
