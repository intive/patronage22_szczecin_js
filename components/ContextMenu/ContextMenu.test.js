/* globals describe, expect, it */
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ContextMenu from './ContextMenu'
import ContextMenuItem from '../ContextMenuItem/ContextMenuItem'

const menuRenderer = () => (
  <ContextMenu id='1'>
    <ContextMenuItem name='setPassword' icon='lock' />
    <ContextMenuItem name='deleteBoard' icon='delete' />
  </ContextMenu>
)

describe('ContextMenu', () => {
  it('should renders correctly ContextMenu', () => {
    const { container } = render(menuRenderer())
    expect(container).toMatchSnapshot()
  })

  it('should open navigation on button click', () => {
    render(menuRenderer())

    expect(screen.getByRole('button')).toHaveAttribute('aria-expanded', 'false')

    userEvent.click(screen.getByText('more_horiz'))

    expect(screen.getByRole('navigation')).toBeVisible()
    expect(screen.getByRole('navigation').classList.contains('active')).toBeTruthy()

    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-haspopup', 'menu')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-expanded', 'true')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-controls', '1')
  })

  it('should open navigation on button click and close when clicked button again', () => {
    render(menuRenderer())

    userEvent.click(screen.getByText('more_horiz'))
    expect(screen.getByRole('navigation')).toBeVisible()
    expect(
      screen.getByRole('navigation').classList.contains('active')
    ).toBeTruthy()

    userEvent.click(screen.getByText('more_horiz'))
    expect(screen.queryByRole('navigation')).toBeNull()

    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-haspopup', 'menu')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-expanded', 'false')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-controls', '1')
  })

  it('should open navigation on button click and close when clicked outside', () => {
    render(
      <>
        {menuRenderer()}
        <div>outsideDiv</div>
      </>
    )

    userEvent.click(screen.getByText('more_horiz'))
    expect(screen.getByRole('navigation')).toBeVisible()
    expect(screen.getByRole('navigation').classList.contains('active')).toBeTruthy()

    userEvent.click(screen.getByText('outsideDiv'))
    expect(screen.queryByRole('navigation')).toBeNull()

    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-haspopup', 'menu')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-expanded', 'false')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-controls', '1')
  })

  it('should open navigation on button click and close when clicked inside navigation', () => {
    render(
      <>
        {menuRenderer()}
        <div>outsideDiv</div>
      </>
    )

    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.getByRole('navigation'))

    expect(screen.queryByRole('navigation')).toBeNull()

    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-haspopup', 'menu')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-expanded', 'false')
    expect(screen.getAllByRole('button')[0]).toHaveAttribute('aria-controls', '1')
  })
})
