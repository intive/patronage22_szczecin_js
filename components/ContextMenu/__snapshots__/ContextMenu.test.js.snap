// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu should renders correctly ContextMenu 1`] = `
<div>
  <div
    class="sc-gsDKAQ kabCuV"
  >
    <button
      aria-controls="1"
      aria-expanded="false"
      aria-haspopup="menu"
      class="sc-dkPtRN fTfaNe"
    >
      <span
        class="material-icons"
      >
        more_horiz
      </span>
    </button>
    <nav
      class="sc-hKwDye chYRkN"
      id="1"
    >
      <ul
        class="sc-bdvvtL bPqLpk"
      >
        <li
          class="sc-eCImPb ihRWfz"
        >
          <button
            class="sc-jRQBWg IsWJg"
          >
            <span
              class="material-icons sc-gKclnd ezggst"
            >
              lock
            </span>
            setPassword
          </button>
        </li>
        <li
          class="sc-eCImPb ihRWfz"
        >
          <button
            class="sc-jRQBWg IsWJg"
          >
            <span
              class="material-icons sc-gKclnd ezggst"
            >
              delete
            </span>
            deleteBoard
          </button>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
